src/passkey_child/passkey_child.c

```diff
@@ -78,6 +78,12 @@ int main(int argc, const char *argv[])
             ERROR("Authentication error.\n");
             goto done;
         }
+    } else if (data.action == ACTION_GET_ASSERT) {
+        ret = get_assert_data(&data);
```

Comment:
Inside the function you assign stuff to data that is never used and never freed. That seems as a weird design to me. I suspect that it will be the same in other branches.

Reply:
Can you be more specific by setting an example? Unless you are referring to `tmp_ctx` I'm having problems to identify the specific problems you are referring to.

Reply:
Ok, looking into that more closely, it is actually worse :-)

1. You pass in data as `ret = get_assert_data(&data);`
2. In `get_assert_data` you assign to various fields, e.g. `ret = get_assert_user_id(tmp_ctx, assert, &data->user_id);`
3. In this case `user_id` is attached to `tmp_ctx` instead of `data` :warning:
4. `tmp_ctx` is not freed :warning:
5. When returning to `main`, `data->user_id` is accessible but never used :warning:

I didn't check every function that is called in `get_assert_data`, but clearly there are issues. If the intention is to fill up the `data` structure then it must be used as the memory context. However, since `data` is not used anywhere in `main` after this cal, then imho it should be treated as input variable only.

Either way, you need to free all allocated memory. Even though it is only short live child process, memory leaks are not symbol of a quality code and should be fixed.

Reply:
Right, I was checking it. `user_id` is attached to `tmp_ctx` and I wasn't freeing this value. This is fixed in the latest version. If you don't like it this way I can attach it to `data` instead.

Apart from that, I've been checking other allocated variables in every branch and I haven't found other non-freed values. Still, I'm following the same approach and I'm attaching the variables to `tmp_ctx`, and then I'm freeing it when I'm done.

Reply:
Attaching data->user_id to tmp_ctx instead of data breaks talloc context hierarchy, it is a bad practice.

I would really just create local variable `unsigned char *user_id;` and use it instead of `data->user_id`.